src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,330 @@
+
+
+class Dimension(SimpleGroupBy, ABC):
+    @abstractmethod
+    def resolve_filter(self, raw_filter: List[str]) -> List[DataCategory]:
+        """
+        Based on the input filter, map it back to the clickhouse representation
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def map_row(self, row: Dict[str, Any]) -> None:
+        """
+        map clickhouse representation back to presentation
+        """
+        raise NotImplementedError()
+
+
+class CategoryDimension(Dimension):
+    def resolve_filter(self, raw_filter: List[str]) -> List[DataCategory]:
+        resolved_categories = set()
+        for category in raw_filter:
+            # combine DEFAULT, ERROR, and SECURITY as errors.
+            # see relay: py/sentry_relay/consts.py and relay-cabi/include/relay.h
+            if DataCategory.parse(category) == DataCategory.ERROR:
+                resolved_categories.update(DataCategory.error_categories())
+            else:
+                resolved_categories.add(DataCategory.parse(category))
+        if DataCategory.ATTACHMENT in resolved_categories and len(resolved_categories) > 1:
+            raise InvalidQuery("if filtering by attachment no other category may be present")
+        return list(resolved_categories)
+
+    def map_row(self, row: Dict[str, Any]) -> None:
+        if "category" in row:
+            category = (
+                DataCategory.ERROR
+                if row["category"] in DataCategory.error_categories()
+                else DataCategory(row["category"])
+            )
+            row["category"] = category.api_name()
+
+
+class OutcomeDimension(Dimension):
+    def resolve_filter(self, raw_filter: List[str]) -> List[Outcome]:
+        return [Outcome.parse(o) for o in raw_filter]
+
+    def map_row(self, row: Dict[str, Any]) -> None:
+        if "outcome" in row:
+            row["outcome"] = Outcome(row["outcome"]).api_name()
+
+
+class ReasonDimension(Dimension):
+    def resolve_filter(self, raw_filter: List[str]) -> List[str]:
+        return [
+            "smart_rate_limit" if reason == "spike_protection" else reason for reason in raw_filter
+        ]
+
+    def map_row(self, row: Dict[str, Any]) -> None:
+        if "reason" in row:
+            row["reason"] = (
+                "spike_protection" if row["reason"] == "smart_rate_limit" else row["reason"]
+            )  # return spike protection to be consistent with naming in other places
+
+
+COLUMN_MAP = {
+    "sum(quantity)": QuantityField(),
+    "sum(times_seen)": TimesSeenField(),
+}
+
+DIMENSION_MAP = {
+    "outcome": OutcomeDimension("outcome"),
+    "category": CategoryDimension("category"),
+    "reason": ReasonDimension("reason"),
+}
+
+GROUPBY_MAP = {
+    **DIMENSION_MAP,
+    "project": SimpleGroupBy("project_id", "project"),
+}
+
+TS_COL = "time"
+
+
+def get_filter(
+    query: QueryDict, params: Dict[Any, Any]
+) -> Tuple[List[Tuple[str, str, List[Any]]], Dict[str, Any]]:
```

Comment:
This is where a NamedTuple for a single Condition would clean this up a bit. Or if returning conditions and filter keys in a tuple happens a lot, wrapping that in a NamedTuple.

Reply:
Tried to make it a namedtuple, but looks like our snuba json serializer likes to convert namedtuples to dicts, which made snuba unhappy about that. Figured it's best not to jump down that rabbit hole so just using a regular tuple type variable, definitely looks cleaner. thanks!